### ISSUE TITLE
libfuse: update to 3.x for speeed at next core

### ISSUE THREAD

**Issue by Artoria2e5** [AOSC-Dev/aosc-os-abbs#2033]
libfuse has been at version 3 for quite a long time. A bunch of defaults are changed, and the SONAME was bumped up, so we are looking at an ABI break here.

libfuse3 is designed to coexist with libfuse2 with some soname magic. Arch Linux is already doing that.

See the changelog at https://github.com/libfuse/libfuse/blob/master/ChangeLog.rst#libfuse-300-2016-12-08.

### Patching packages
If we want some filesystems to be really, really fast, we can try porting them to the new libfuse (and yelling at the upstream for it, of course). A few things we should watch for are:

* libfuse3 always sets an option that assumes setu/gid to be dropped by the fs itself on chmod, trunc, and write. For security that needs to be done.
* API changes on init. We only need to change the function signature since we aren't using the new config info.
* API changes on handlers. As with init, the new `fuse_file_info` struct is not used and we only need to change the signature. The removed handlers can just be removed.
* build system: fuse 3.x has a new package name `fuse3`, so detection and emission of separate `#define`s for `#ifdef` should be trivial.

I accidentally backported some of libfuse3's improvements into libfuse2 when I was working on our [ntfs-3g](https://github.com/AOSC-Dev/ntfs-3g) patchset. As a result the internal libfuse for ntfs-3g is quite a bit faster now. We can, of course, patch *that* back into our fuse2 and make something (i don't know) fast when it wants big writes.

**Comment by jyhi** [comment]
```
$ apt rdepends fuse
fuse
Reverse Depends:
  Depends: fuseiso (>= 2.9.3-4)
  Depends: sshfs (>= 2.9.7)
  Depends: libvshadow (>= 2.9.8)
  Depends: admin-base
  Depends: gvfs (>= 2.9.8-1)
  Depends: curlftpfs (>= 2.9.7)
  Depends: dislocker (>= 2.9.8)
  Depends: grub (>= 2.9.8-1)
  Depends: sra-tools (>= 2.9.7)
  Depends: fossil (>= 2.9.8)
  Depends: ifuse (>= 2.9.7)
  Depends: llfuse (>= 2.9.8-1)
  Depends: cromfs (>= 2.9.7)
  Depends: xrdp (>= 2.9.8-1)
  Depends: veracrypt (>= 2.9.8)
  Depends: ostree (>= 2.9.8)
  Depends: cryfs (>= 2.9.8)
  Depends: ceph (>= 2.9.8)
  Depends: flatpak (>= 2.9.8)
  Depends: libguestfs (>= 2.9.8-1)
  Depends: fuse-python (>= 2.9.7)
  Depends: afflib (>= 2.9.8-1)
  Depends: wimlib (>= 2.9.8)
  Depends: avfs (>= 2.9.7)
  Depends: ntfs-3g (>= 2.9.7)
  Depends: compiz (>= 2.9.7)
  Depends: afflib (>= 2.9.8)
  Depends: compiz (>= 2.9.8-1)
  Depends: fuse-exfat (>= 2.9.7)
  Depends: llfuse (>= 2.9.8)
  Depends: cryfs (>= 2.9.8-1)
  Depends: xrdp (>= 2.9.8)
  Depends: afpfs-ng (>= 2.9.7)
  Depends: lxcfs (>= 2.9.8)
  Depends: darling-dmg (>= 2.9.8)
  Depends: ostree (>= 2.9.8-1)
  Depends: encfs (>= 2.9.8)
  Depends: ostree (>= 2.9.8-1)
  Depends: avfs (>= 2.9.8-1)
  Depends: open-vm-tools (>= 2.9.8-1)
  Depends: xrdp (>= 2.9.8-1)
  Depends: libguestfs (>= 2.9.8)
  Depends: llfuse (>= 2.9.8-1)
  Depends: open-vm-tools (>= 2.9.8-1)
```

**Comment by jyhi** [comment]
I use fuse 3 in my final year project, but I don't know the difference between fuse 2 and 3. As API changes may lead to design decision, will it be better if we first contact authors of packages depending on fuse 2?

**Comment by MingcongBai** [comment]
@lmy441900 The current assumption is that, since other distributions (Arch Linux, Debian, Fedora, ...) are keeping both FUSE 2.x/3.x, and only a small collection of packages depends on the 3.x library, that there is some sort of incompatibility...

**Comment by Artoria2e5** [comment]
There is. That's why I mentioned porting and the API changes.

**Comment by Artoria2e5** [comment]
Done porting wimlib: https://github.com/Artoria2e5/wimlib-fuse3/commit/9324650c18044085a3f8127b5ea1cb61799ffec9

**Comment by MingcongBai** [comment]
Addressed with 181638887c3d81f2cbdc9aaf9a7dfdacf5bd970e (`fuse-3` introduction). However, upstream adoption is still ongoing - there is no rush and, if anyone would like to pick this up, please re-open this issue.
